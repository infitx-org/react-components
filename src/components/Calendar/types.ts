export type PossibleDate = Date | undefined;

export type PossibleDay = string | undefined;

export type Matrix = PossibleDay[][];

export type DisabledDays = (day: Date) => boolean;

export type DateRange = [PossibleDate, PossibleDate];
