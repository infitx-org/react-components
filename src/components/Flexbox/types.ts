export enum TopBottom {
  "top" = "flex-start",
  "bottom" = "flex-end",
}
export enum LeftRight {
  "left" = "flex-start",
  "right" = "flex-end",
}

export enum Mappers {
  "top" = "flex-start",
  "bottom" = "flex-end",
  "left" = "flex-start",
  "right" = "flex-end",
}

export type AlignItems =
  | "flex-start"
  | "flex-end"
  | "center"
  | "baseline"
  | "stretch";

export type JustifyContent =
  | "flex-start"
  | "flex-end"
  | "center"
  | "space-between"
  | "space-around"
  | "space-evenly";

export type AlignWithMap<T extends string> = `${AlignItems}` | `${T}`;
export type JustifyWithMap<T extends string> = `${JustifyContent}` | `${T}`;
