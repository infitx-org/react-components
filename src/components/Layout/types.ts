export enum AlignMap {
  "top" = "flex-start",
  "bottom" = "flex-end",
}

export enum JustifyMap {
  "left" = "flex-start",
  "right" = "flex-end",
}

export type AlignItems =
  | "flex-start"
  | "flex-end"
  | "center"
  | "baseline"
  | "stretch";

export type JustifyContent =
  | "flex-start"
  | "flex-end"
  | "center"
  | "space-between"
  | "space-around"
  | "space-evenly";

export type AlignWithMap = `${AlignItems | keyof typeof AlignMap}`;
export type JustifyWithMap = `${JustifyContent | keyof typeof JustifyMap}`;
