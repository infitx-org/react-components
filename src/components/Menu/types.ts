export interface MenuSectionProps {
  label: string;
  hidden?: boolean;
  icon?: React.ReactElement<React.SVGProps<SVGSVGElement>>;
  fill?: string;
  size?: number;
  disabled?: boolean;
  children: MenuItemElement | MenuItemElement[];
}

interface BaseMenuItemProps {
  label: string;
  active?: boolean;
  disabled?: boolean;
  hidden?: boolean;
  back?: boolean;
  icon?: React.ReactElement<React.SVGProps<SVGSVGElement>>;
  fill?: string;
  size?: number;
  // prop partial is not used directly by the MenuItem but it is used by the Menu component
  partial?: boolean;
  children?: MenuItemElement | MenuElement[];
}

interface BaseMenuItemPropsWithTo extends BaseMenuItemProps {
  to: string;
  path?: string;
}
interface BaseMenuItemPropsWithPath extends BaseMenuItemProps {
  to?: string;
  path: string;
}

export type MenuItemProps = BaseMenuItemPropsWithTo | BaseMenuItemPropsWithPath;

export type MenuItemElement = React.ReactElement<MenuItemProps>;
export type MenuSectionElement = React.ReactElement<MenuSectionProps>;
export type MenuElement = MenuItemElement | MenuSectionElement;
