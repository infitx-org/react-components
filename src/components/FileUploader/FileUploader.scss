@use "../../scss/base/colors.scss";
@use "../../scss/base/transitions.scss" as *;
@use "../../scss/components/field.scss";

.rc-fileuploader__input {
  opacity: 0;
  z-index: -1;
  position: absolute;
}

.rc-fileuploader {
  @include field.rc-field-input;

  flex: 2 1 auto;
  position: relative;
  height: 20px;
  vertical-align: middle;
  line-height: 20px;
  pointer-events: none;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  color: colors.$dark-gray;
}
