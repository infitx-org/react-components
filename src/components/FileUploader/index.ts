export { default, FileUploaderProps } from "./FileUploader";
