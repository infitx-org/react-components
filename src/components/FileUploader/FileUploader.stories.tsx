import FileUploader from "./FileUploader";

export default {
  title: "Components/FileUploader",
};

/* eslint-disable no-console */

const file = new File([new Blob(["test"], { type: "text/plain" })], "test");
const Template = (args) => <FileUploader {...args} />;

export const Default = Template.bind({});
Default.args = {
  size: "large",
  file,
  className: undefined,
  placeholder: "Choose a file",
  required: false,
  invalid: false,
  pending: false,
  onChange: console.log,
};

export const SelectedFile = Template.bind({});
SelectedFile.args = {
  file,
};
export const SelectedFileName = Template.bind({});
SelectedFileName.args = {
  fileName: "myfile.dat",
};

export const Disabled = Template.bind({});
Disabled.args = {
  disabled: true,
};

export const Placeholder = Template.bind({});
Placeholder.args = {
  placeholder: "Pick a file",
};

export const Pending = Template.bind({});
Pending.args = {
  pending: true,
};

export const Required = Template.bind({});
Required.args = {
  required: true,
};

export const Invalid = Template.bind({});
Invalid.args = {
  invalid: true,
};

export const Small = Template.bind({});
Small.args = {
  size: "small",
};

export const Medium = Template.bind({});
Medium.args = {
  size: "medium",
};

export const Large = Template.bind({});
Large.args = {
  size: "large",
};
