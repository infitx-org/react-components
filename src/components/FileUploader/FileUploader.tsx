import React from "react";
import classnames from "classnames";
import { Kind, InputSize, KeyCode } from "types";
import readFileAsText from "@modusbox/ts-utils/lib/file/readFileAsText";
import readFileAsBase64 from "@modusbox/ts-utils/lib/file/readFileAsBase64";
import Field, {
  Loader,
  Placeholder,
  InvalidIcon,
  ValidationProps,
} from "components/Field";
import Button from "components/Button";
import { BaseInput } from "components/shared/types";
import "./FileUploader.scss";

export interface BaseFileUploaderProps extends BaseInput {
  kind?: `${Kind}`;
  size?: `${InputSize}`;
  className?: string;
  parseAs?: "text" | "base64";
  placeholder?: string;
  file?: File;
  fileName?: string;
  required?: boolean;
  invalid?: boolean;
  pending?: boolean;
  onChange?: (content?: string) => void;
}

export type FileUploaderProps = BaseFileUploaderProps & ValidationProps;

export default React.forwardRef(function FileUploader(
  {
    kind = Kind.Primary,
    size = InputSize.Large,
    className,
    parseAs = "text",
    placeholder,
    file,
    fileName,
    required,
    invalid,
    pending,
    onChange,
    ...props
  }: FileUploaderProps,
  forwardedRef: React.ForwardedRef<HTMLDivElement>
): JSX.Element {
  const inputRef = React.useRef<HTMLInputElement>(null);
  const [selectedFile, setFile] = React.useState<File | undefined>(file);
  const [selectedFileName, setFileName] = React.useState<string | undefined>(
    fileName || file?.name
  );
  const [open, setOpen] = React.useState(false);
  const [focused, setFocused] = React.useState(false);

  React.useEffect(() => {
    setFile(file);
  }, [file]);

  React.useEffect(() => {
    if (!selectedFile && selectedFileName) {
      setFileName(fileName);
    }
  }, [fileName]);

  function enter() {
    setFocused(true);
    setOpen(true);
    inputRef.current?.focus();
  }

  function leave() {
    setFocused(false);
    setOpen(false);
  }

  function onChooseFileButtonClick(e: React.MouseEvent<HTMLButtonElement>) {
    e.stopPropagation();
    inputRef.current?.click();
  }
  function onRemoveFileButtonClick() {
    setFile(undefined);
    setFileName(undefined);
    onChange?.(undefined);
  }

  async function onFileChange(e: React.ChangeEvent<HTMLInputElement>) {
    const inputFile = e.target.files?.item(0);
    if (!inputFile) {
      return;
    }
    let parse;
    if (parseAs === "text") {
      parse = readFileAsText;
    } else {
      parse = readFileAsBase64;
    }
    const content = await parse(inputFile);
    setFile(inputFile);
    setFileName(inputFile.name);
    onChange?.(content as string);
    inputRef.current?.focus();
  }

  function onFocus(e: React.FocusEvent<HTMLInputElement>) {
    if (!focused) {
      props.onFocus?.(e);
      enter();
    }
  }

  function onBlur(e: React.FocusEvent<HTMLInputElement>) {
    if (!focused) {
      props.onBlur?.(e);
    }
  }

  function onFieldClick(e: React.MouseEvent<HTMLDivElement>): void {
    if (e.target !== inputRef.current) {
      e.preventDefault();
      e.stopPropagation();
      if (open) {
        leave();
      } else {
        enter();
      }
    }
  }

  function onKeyDown(e: React.KeyboardEvent<HTMLInputElement>) {
    const { keyCode } = e;

    if (keyCode === KeyCode.Tab) {
      leave();
      return;
    }
    if (keyCode === KeyCode.Return) {
      e.preventDefault();
      if (selectedFileName) {
        onRemoveFileButtonClick();
      } else {
        inputRef.current?.click();
      }
    }
  }

  const button = (
    <Button
      noFill={!focused}
      tabIndex={-1}
      size="extra-small"
      label={selectedFileName ? "Remove File" : "Choose File"}
      onClick={
        selectedFileName ? onRemoveFileButtonClick : onChooseFileButtonClick
      }
      kind={selectedFileName ? "danger" : "primary"}
    />
  );

  const fileNameClassName = classnames([
    "rc-fileuploader",
    `rc-fileuploader--${size}`,
  ]);

  return (
    <Field
      className={className}
      kind={kind}
      size={size}
      required={required && selectedFile === undefined}
      invalid={invalid}
      disabled={props.disabled}
      focused={focused}
      onClick={onFieldClick}
      onClickOutside={leave}
      ref={forwardedRef}
      validation={props.validation}
    >
      {placeholder && (
        <Placeholder
          label={placeholder}
          active={!!selectedFile || focused}
          size={size}
        />
      )}
      <input
        {...props}
        className="rc-fileuploader__input"
        type="file"
        ref={inputRef}
        onChange={onFileChange}
        onFocus={onFocus}
        onBlur={onBlur}
        onKeyDown={onKeyDown}
      />
      <div className={fileNameClassName}>{selectedFileName}</div>
      {pending && <Loader size={size} />}
      {invalid && <InvalidIcon size={size} />}
      {button}
    </Field>
  );
});
