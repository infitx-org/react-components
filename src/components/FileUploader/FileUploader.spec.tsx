import { render, fireEvent, waitFor, act } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import userEvent from "@testing-library/user-event";
import { Kind, InputSize } from "types";
import FileUploader from "./FileUploader";

const testFile = new File(
  [new Blob(["test"], { type: "text/plain" })],
  "test.dat"
);
const commonProps = {
  onChange: jest.fn(),
};

function getInput(container: HTMLElement): HTMLInputElement {
  return container.querySelector("input[type='file']") as HTMLInputElement;
}

function getButton(container: HTMLElement): HTMLInputElement {
  return container.querySelector("button") as HTMLInputElement;
}

function getFileName(container: HTMLElement): HTMLInputElement {
  return container.querySelector(".rc-fileuploader") as HTMLInputElement;
}

describe("tests the FileUploader props", () => {
  it("renders the FileUploader", () => {
    const { container } = render(<FileUploader {...commonProps} />);
    expect(container.querySelector(".rc-fileuploader")).toBeTruthy();
    expect(container.querySelectorAll('input[type="file"]')).toHaveLength(1);
  });

  it("renders the file", () => {
    const { container } = render(
      <FileUploader {...commonProps} file={testFile} />
    );
    const fileName = getFileName(container);
    expect(fileName).toHaveTextContent(testFile.name);
  });

  it("renders no filename when file is not specified", () => {
    const { container } = render(<FileUploader {...commonProps} />);
    const fileName = getFileName(container);
    expect(fileName).toHaveTextContent("");
  });

  it("renders the placeholder", () => {
    const { container } = render(
      <FileUploader {...commonProps} placeholder="test" />
    );
    expect(container.querySelector(".placeholder")).toBeTruthy();
  });

  it("renders the classname", () => {
    const { container } = render(
      <FileUploader {...commonProps} className="test" />
    );
    expect(container.querySelector(".test")).toBeTruthy();
  });

  it("renders as disabled", () => {
    const { container } = render(<FileUploader {...commonProps} disabled />);
    const input = getInput(container);
    expect(input.disabled).toBeTruthy();
  });

  it("renders as pending", () => {
    const { container } = render(<FileUploader {...commonProps} pending />);
    expect(container.querySelector(".rc-spinner")).toBeTruthy();
  });

  it("renders as required", () => {
    const { container } = render(<FileUploader {...commonProps} required />);
    expect(container.querySelector(".rc-field--required")).toBeTruthy();
  });

  it("renders as invalid", () => {
    const { container } = render(<FileUploader {...commonProps} invalid />);
    expect(container.querySelector(".rc-field--invalid")).toBeTruthy();
    expect(container.querySelector(".rc-invalid-icon")).toBeTruthy();
  });

  it("renders the small, medium, large sizes", () => {
    Object.values(InputSize).forEach((size) => {
      const { container } = render(
        <FileUploader {...commonProps} size={size} />
      );
      expect(container.querySelector(`.rc-field--${size}`)).toBeTruthy();
    });
  });

  it("renders all the kinds", () => {
    Object.values(Kind).forEach((kind) => {
      const { container } = render(
        <FileUploader {...commonProps} kind={kind} />
      );
      expect(container.querySelector(`.rc-field--${kind}`)).toBeTruthy();
    });
  });

  it("triggers onFocus when focusing", () => {
    const mockFn = jest.fn();
    const { container } = render(
      <FileUploader {...commonProps} onFocus={mockFn} />
    );
    fireEvent.focus(getInput(container));
    expect(mockFn).toHaveBeenCalled();
  });

  it("triggers onFocus when clicking", () => {
    const mockFn = jest.fn();
    const { container } = render(
      <FileUploader {...commonProps} onFocus={mockFn} />
    );
    userEvent.click(getInput(container));
    expect(mockFn).toHaveBeenCalled();
  });

  it("triggers onFocus when tabbing", () => {
    const mockFn = jest.fn();
    render(<FileUploader {...commonProps} onFocus={mockFn} />);
    userEvent.tab();
    expect(mockFn).toHaveBeenCalled();
  });

  it("triggers onBlur when clicking out", () => {
    const mockFn = jest.fn();
    const { container } = render(
      <FileUploader {...commonProps} onBlur={mockFn} />
    );
    userEvent.click(getInput(container));
    userEvent.click(document.body);
    expect(mockFn).toHaveBeenCalled();
  });

  // it("triggers onBlur when tabbing away", () => {
  //   const mockFn = jest.fn();
  //   const { container } = render(<FileUploader {...commonProps} onBlur={mockFn} />);
  //   userEvent.click(getInput(container));
  //   userEvent.tab();
  //   expect(mockFn).toHaveBeenCalled();
  // });

  it("trigger onChange and selects the file", async () => {
    const mockFn = jest.fn();
    const { container } = render(<FileUploader onChange={mockFn} />);
    await waitFor(() => userEvent.upload(getInput(container), testFile));
    const input = await waitFor(() => getInput(container));
    expect(input.files?.[0]?.name).toBe(testFile.name);
    expect(getFileName(container)).toHaveTextContent(testFile.name);
    expect(mockFn).toHaveBeenCalledWith("test");
  });

  it("trigger onChange when removing the selected file", async () => {
    const mockFn = jest.fn();
    const { container } = render(
      <FileUploader onChange={mockFn} file={testFile} />
    );
    fireEvent.click(getButton(container) as Element);
    expect(getFileName(container)).toHaveTextContent("");
    expect(mockFn).toHaveBeenCalledWith(undefined);
  });
});

// Snapshot testing
it("renders the FileUploader correctly when multiple props are set", () => {
  const { container } = render(
    <FileUploader {...commonProps} placeholder="test" />
  );
  expect(container).toMatchSnapshot();
});
