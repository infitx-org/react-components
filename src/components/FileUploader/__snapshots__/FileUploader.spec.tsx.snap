// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the FileUploader correctly when multiple props are set 1`] = `
<div>
  <div
    class="rc-field rc-field--primary rc-field--large"
    role="presentation"
  >
    <label
      class="placeholder placeholder--large"
    >
      test
    </label>
    <input
      class="rc-fileuploader__input"
      type="file"
    />
    <div
      class="rc-fileuploader rc-fileuploader--large"
    />
    <button
      class="rc-icon-button rc-icon-button--default rc-fileuploader__button"
      style="height: 18px; width: 18px;"
      type="button"
    >
      <svg
        class="rc-icon-button__icon"
        fill="#000"
        height="18"
        stroke="transparent"
        style="fill: inherit;"
        width="18"
      />
    </button>
  </div>
</div>
`;
